File: cf/contest/1362/d/main.py
import sys

sys.setrecursionlimit(10 ** 8)
ini = lambda: int(sys.stdin.readline())
inm = lambda: map(int, sys.stdin.readline().split())
inl = lambda: list(inm())
ins = lambda: sys.stdin.readline().rstrip()


def main():
    N, M = inm()
    adj = [[] for _ in range(N)]
    for i in range(M):
        a, b = inm()
        a -= 1
        b -= 1
        assert a != b
        adj[a].append(b)
        adj[b].append(a)
    topic = inl()
    assert len(topic) == N
    cntr = 0
    order = [None] * N

    def solve(cur):
        nonlocal cntr
        t = topic[cur]
        covered = set()
        for nei in adj[cur]:
            if topic[nei] == t:
                return False
            if topic[nei] < t:
                covered.add(topic[nei])
        if len(covered) < t - 1:
            # print("cannot place: {} topic  covered {}".format(cur, t, covered))
            return False
        if order[cur] is None:
            order[cur] = cntr
            cntr += 1
        return True

    pres = sorted([(topic[i], i) for i in range(N)])
    for i in range(N):
        _, j = pres[i]
        if order[j] is None:
            if not solve(j):
                print(-1)
                return
    ans = [None] * N
    for i in range(N):
        ans[order[i]] = i + 1
    print(*ans)


main()

File: cf/contest/1362/d/main.cpp
//#define DEBUGGING  // Enables DEBUG macro.
#include <bits/stdc++.h>
using namespace std;
using i64 = long long;
using u64 = unsigned long long;
#define REP(i, n) for (int i = 0; (i64)(i) < (i64)(n); ++i)

#ifndef DEBUGGING
#define DEBUG(...)
#else
template <typename T>
void debug(T value) {
  std::cerr << value;
}
template <typename T, typename... Ts>
void debug(T value, Ts... args) {
  std::cerr << value << ", ";
  debug(args...);
}
#define DEBUG(...)                     \
  do {                                 \
    cerr << " (L" << __LINE__ << ") "; \
    cerr << #__VA_ARGS__ << ": ";      \
    debug(__VA_ARGS__);                \
    cerr << endl;                      \
  } while (0)
#endif

int main() {
  ios_base::sync_with_stdio(false);
  cin.tie(nullptr);
  i64 n, m;
  cin >> n >> m;
  vector<vector<i64>> adj(n);
  REP(i, m) {
    i64 a, b;
    cin >> a >> b;
    a--;
    b--;
    assert(a != b);
    adj[a].emplace_back(b);
    adj[b].emplace_back(a);
  }
  vector<i64> topic(n);
  for (auto& x : topic) cin >> x;

  int cntr = 0;
  vector<int> ord(n, -1);

  auto solve = [&](auto rec, int cur) -> void {
    i64 t = topic[cur];
    vector<int> neis;
    set<int> covered;
    for (auto nei : adj[cur]) {
      if (topic[nei] == t) {
        cout << "-1" << endl;
        exit(0);
      }
      if (topic[nei] < t) {
        covered.insert(topic[nei]);
      }
    }
    if (covered.size() != t - 1) {
      cout << -1 << endl;
      exit(0);
    }
    if (ord[cur] < 0) {
      ord[cur] = cntr++;
    }
    // for (auto nei : neis) {
    //   if (ord[nei] < 0) {
    //     rec(rec, nei);
    //   }
    // }
  };

  vector<tuple<i64, int>> pres(n);
  REP(i, n) { pres[i] = {topic[i], i}; }
  sort(pres.begin(), pres.end());
  REP(i, n) {
    int j = get<1>(pres[i]);
    if (ord[j] < 0) {
      solve(solve, j);
    }
  }
  assert(cntr == n);

  vector<int> ans(n);
  REP(i, n) { ans[ord[i]] = i + 1; }
  REP(i, n) {
    if (i > 0) cout << ' ';
    cout << ans[i];
  }
  cout << endl;
}
